src/common/compile.ts

```diff
@@ -0,0 +1,493 @@
+  if (
+    options.objective === "bytes" ||
+    packers.length < 1 ||
+    isError(bestUnpacked.result)
+  )
+    return bestUnpacked;
+  function packer(code: string | null): string | null {
+    if (code === null) return null;
+    if ([...code].map((x) => x.charCodeAt(0)).some((x) => x > 127)) return null;
+    return packers
+      .map((x) => x(code))
+      .reduce((a, b) => (obj(a) < obj(b) ? a : b));
+  }
+  const bestForPacking = compileVariantNoPacking(
+    program,
+    {
+      ...options,
+      objective: (x) => charLength(packer(x)),
+    },
+    language
+  );
+  if (isError(bestForPacking.result)) return bestUnpacked;
+  const packed = packer(bestForPacking.result);
+  if (packed != null && obj(packed) < obj(bestUnpacked.result)) {
+    return { ...bestForPacking, result: packed };
+  }
+  return bestUnpacked;
+}
+
+function shorterBy(
+  obj: (x: string | null) => number
+): (a: CompilationResult, b: CompilationResult) => CompilationResult {
+  return (a, b) =>
+    isError(a.result)
+      ? b
+      : isError(b.result)
+      ? a
+      : obj(a.result) < obj(b.result)
+      ? a
+      : b;
+}
+
+interface SearchState {
+  program: Program;
+  startPhase: number;
+  length: number;
+  warnings: Error[];
+  history: [number, string][];
+}
+
+export function compileVariantNoPacking(
+  program: Program,
+  options: CompilationOptions,
+  language: Language
+): CompilationResult {
+  const phases = language.phases;
+  if (options.level === "nogolf" || options.level === "simple") {
+    try {
+      const warnings: Error[] = [];
+      const addWarning = (x: Error) => warnings.push(x);
+      const plugins = phases
+        .filter(
+          options.level === "nogolf"
+            ? (x) => x.mode === "required"
+            : (x) => x.mode !== "search"
+        )
+        .flatMap((x) => x.plugins);
+      const [res, counts] = applyAllToAllAndGetCounts(
+        program,
+        addWarning,
+        options,
+        ...plugins.map((x) => x.visit)
+      );
+      return compilationResult(
+        language.name,
+        emit(language, res, addWarning, options),
+        plugins.map((y, i) => [counts[i], y.name]),
+        warnings
+      );
+    } catch (e) {
+      if (isError(e)) {
+        return compilationResult(language.name, e);
+      }
+      throw e;
+    }
+  }
+  const obj = getObjectiveFunc(options);
+  function finish(
+    prog: Program,
+    addWarning: AddWarning,
+    startPhase = 0
+  ): [string, [number, string][]] {
+    const finishingPlugins = phases
+      .slice(startPhase)
+      .filter((x) => x.mode !== "search")
+      .flatMap((x) => x.plugins);
+    const [resProg, counts] = applyAllToAllAndGetCounts(
+      prog,
+      addWarning,
+      options,
+      ...finishingPlugins.map((x) => x.visit)
+    );
+    return [
+      emit(language, resProg, addWarning, options),
+      finishingPlugins.map((x, i) => [counts[i], x.name]),
+    ];
+  }
+  let shortestSoFar: SearchState;
```

Comment:
If you type this as `let shortestSoFar: SearchState | undefined;`, then you don't need any of the `shortestSoFar!` assertions.

Reply:
I tried many different version around this and I was encountering some very weird behaviours that I couldn't understand and considered them possible bugs. This ugly thing was the only thing that worked for me. :(

Reply:
Never mind, your suggestion apparently works.
I guess, I was only trying `let shortestSoFar: SearchState | undefined = undefined;` which doesn't work as typescript narrows the variable type to just `undefined` in the section after the `while`. Like it doesn't understand that the `enqueue` closes over the `shortestSoFar` variable or something....